File: cheat_sheet/go/2_extend_feature/internal/domain/auto_reply/trigger_validation.go
package auto_reply

import (
	"encoding/json"
	"sort"
	"strings"
	"time"

	"github.com/chatbotgang/workshop/internal/domain/organization"
)

// WebhookEvent represents a generic webhook event for trigger validation
type WebhookEvent struct {
	EventType   string               `json:"event_type"`            // "message" (only type we handle for now)
	MessageText *string              `json:"message_text"`          // For keyword matching
	Timestamp   time.Time            `json:"timestamp"`             // For time-based triggers
	ChannelType organization.BotType `json:"channel_type"`          // LINE, FB, IG
	IGStoryID   *string              `json:"ig_story_id,omitempty"` // IG Story ID for story-specific triggers
}

// AutoReplyTriggerSetting represents the merged domain model of AutoReply and WebhookTriggerSetting
// This provides easy access to both trigger configuration and priority information
type AutoReplyTriggerSetting struct {
	// AutoReply fields
	AutoReplyID        int                `json:"auto_reply_id"`
	AutoReplyName      string             `json:"auto_reply_name"`
	AutoReplyStatus    AutoReplyStatus    `json:"auto_reply_status"`
	AutoReplyEventType AutoReplyEventType `json:"auto_reply_event_type"`
	AutoReplyPriority  int                `json:"auto_reply_priority"` // Key field for sorting
	Keywords           []string           `json:"keywords,omitempty"`

	// WebhookTriggerSetting fields
	WebhookTriggerID        int                         `json:"webhook_trigger_id"`
	BotID                   int                         `json:"bot_id"`
	Enable                  bool                        `json:"enable"`
	WebhookTriggerName      string                      `json:"webhook_trigger_name"`
	EventType               WebhookTriggerEventType     `json:"event_type"`
	TriggerCode             *string                     `json:"trigger_code,omitempty"`
	TriggerScheduleType     *WebhookTriggerScheduleType `json:"trigger_schedule_type,omitempty"`
	TriggerScheduleSettings json.RawMessage             `json:"trigger_schedule_settings,omitempty"`
	IGStoryIDs              []string                    `json:"ig_story_ids,omitempty"` // IG Story IDs for story-specific triggers
	Archived                bool                        `json:"archived"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// IsActive returns true if both AutoReply and WebhookTriggerSetting are active
func (arts *AutoReplyTriggerSetting) IsActive() bool {
	return arts.AutoReplyStatus == AutoReplyStatusActive &&
		arts.Enable &&
		!arts.Archived
}

// IsKeywordTrigger returns true if this is a keyword-based trigger (not IG Story-specific)
func (arts *AutoReplyTriggerSetting) IsKeywordTrigger() bool {
	// Must have keywords and NOT be IG Story-specific
	hasKeywords := len(arts.Keywords) > 0 || (arts.TriggerCode != nil && *arts.TriggerCode != "")
	return hasKeywords && len(arts.IGStoryIDs) == 0
}

// IsGeneralTimeTrigger returns true if this is a general time-based trigger (not IG Story-specific)
func (arts *AutoReplyTriggerSetting) IsGeneralTimeTrigger() bool {
	return arts.EventType == EventTypeTime && arts.TriggerScheduleType != nil && len(arts.IGStoryIDs) == 0
}

// IsIGStoryKeywordTrigger returns true if this is an IG Story-specific keyword trigger
func (arts *AutoReplyTriggerSetting) IsIGStoryKeywordTrigger() bool {
	// Must have keywords AND be IG Story-specific
	hasKeywords := len(arts.Keywords) > 0 || (arts.TriggerCode != nil && *arts.TriggerCode != "")
	return hasKeywords && len(arts.IGStoryIDs) > 0
}

// IsIGStoryGeneralTrigger returns true if this is an IG Story-specific general time trigger
func (arts *AutoReplyTriggerSetting) IsIGStoryGeneralTrigger() bool {
	return arts.EventType == EventTypeTime && arts.TriggerScheduleType != nil && len(arts.IGStoryIDs) > 0
}

// AutoReplyChannelSettingAggregate holds all trigger settings for a specific bot/channel
// and manages the validation logic with business rules
type AutoReplyChannelSettingAggregate struct {
	BotID           int                         `json:"bot_id"`
	TriggerSettings []AutoReplyTriggerSetting   `json:"trigger_settings"` // Updated to use merged model
	BusinessHours   []organization.BusinessHour `json:"business_hours"`
	Timezone        string                      `json:"timezone"` // e.g., "Asia/Taipei"
}

// ValidateTrigger validates a webhook event against configured trigger settings
// Returns the highest priority matching trigger setting, or nil if no match
// Priority order for Feature 2 (4-level priority system):
//  1. IG Story Keyword (Priority 1 - highest)
//  2. IG Story General (Priority 2)
//  3. General Keyword (Priority 3)
//  4. General Time-based (Priority 4 - lowest)
//
// Within each level: sorted by auto-reply's priority field (higher number = higher priority)
// For general time triggers: schedule type priority (monthly -> business hour -> non-business hour -> daily)
//
//	then by auto-reply's priority field within the same schedule type
func (a *AutoReplyChannelSettingAggregate) ValidateTrigger(event WebhookEvent) (*AutoReplyTriggerSetting, error) {
	// Only handle MESSAGE events
	if event.EventType != "message" {
		return nil, nil
	}

	// Priority 1: Check IG Story keyword triggers first (highest priority)
	igStoryKeywordMatch, err := a.validateIGStoryKeywordTrigger(event)
	if err != nil {
		return nil, err
	}
	if igStoryKeywordMatch != nil {
		return igStoryKeywordMatch, nil
	}

	// Priority 2: Check IG Story general triggers
	igStoryGeneralMatch, err := a.validateIGStoryGeneralTrigger(event)
	if err != nil {
		return nil, err
	}
	if igStoryGeneralMatch != nil {
		return igStoryGeneralMatch, nil
	}

	// Priority 3: Check general keyword triggers
	keywordMatch, err := a.validateKeywordTrigger(event)
	if err != nil {
		return nil, err
	}
	if keywordMatch != nil {
		return keywordMatch, nil
	}

	// Priority 4: Check general time triggers (lowest priority)
	generalTimeMatch, err := a.validateGeneralTimeTrigger(event)
	if err != nil {
		return nil, err
	}
	if generalTimeMatch != nil {
		return generalTimeMatch, nil
	}

	// No matching triggers found
	return nil, nil
}

// validateKeywordTrigger checks if the event matches any keyword trigger settings
func (a *AutoReplyChannelSettingAggregate) validateKeywordTrigger(event WebhookEvent) (*AutoReplyTriggerSetting, error) {
	if event.MessageText == nil {
		return nil, nil // No message text to match against
	}

	normalizedMessage := normalizeKeyword(*event.MessageText)
	var matchingSettings []AutoReplyTriggerSetting

	// Find all keyword triggers that match the message
	for _, setting := range a.TriggerSettings {
		if !setting.IsActive() {
			continue
		}

		if setting.IsKeywordTrigger() {
			// Check if any of the keywords match
			for _, keyword := range setting.Keywords {
				normalizedKeyword := normalizeKeyword(keyword)
				if normalizedMessage == normalizedKeyword {
					matchingSettings = append(matchingSettings, setting)
					break // Found a match, no need to check other keywords for this setting
				}
			}

			// Also check TriggerCode for backward compatibility
			if setting.TriggerCode != nil && *setting.TriggerCode != "" {
				normalizedTriggerCode := normalizeKeyword(*setting.TriggerCode)
				if normalizedMessage == normalizedTriggerCode {
					matchingSettings = append(matchingSettings, setting)
				}
			}
		}
	}

	if len(matchingSettings) == 0 {
		return nil, nil // No matching keyword triggers
	}

	// Sort by priority and return the highest priority match
	sortedSettings := a.sortKeywordTriggersByPriority(matchingSettings)
	return &sortedSettings[0], nil
}

// normalizeKeyword normalizes keyword for matching (case insensitive, trim spaces)
func normalizeKeyword(keyword string) string {
	// Trim leading and trailing spaces, then convert to lowercase
	return strings.ToLower(strings.TrimSpace(keyword))
}

// validateIGStoryKeywordTrigger checks if the event matches any IG Story keyword trigger settings
func (a *AutoReplyChannelSettingAggregate) validateIGStoryKeywordTrigger(event WebhookEvent) (*AutoReplyTriggerSetting, error) {
	// Must have both message text and IG Story ID
	if event.MessageText == nil || event.IGStoryID == nil {
		return nil, nil
	}

	normalizedMessage := normalizeKeyword(*event.MessageText)
	var matchingSettings []AutoReplyTriggerSetting

	// Find all IG Story keyword triggers that match the message and story ID
	for _, setting := range a.TriggerSettings {
		if !setting.IsActive() {
			continue
		}

		if setting.IsIGStoryKeywordTrigger() {
			// Check if the story ID matches
			if !a.matchesStoryID(*event.IGStoryID, setting.IGStoryIDs) {
				continue
			}

			// Check if any of the keywords match
			for _, keyword := range setting.Keywords {
				normalizedKeyword := normalizeKeyword(keyword)
				if normalizedMessage == normalizedKeyword {
					matchingSettings = append(matchingSettings, setting)
					break // Found a match, no need to check other keywords for this setting
				}
			}

			// Also check TriggerCode for backward compatibility
			if setting.TriggerCode != nil && *setting.TriggerCode != "" {
				normalizedTriggerCode := normalizeKeyword(*setting.TriggerCode)
				if normalizedMessage == normalizedTriggerCode {
					matchingSettings = append(matchingSettings, setting)
				}
			}
		}
	}

	if len(matchingSettings) == 0 {
		return nil, nil // No matching IG Story keyword triggers
	}

	// Sort by priority and return the highest priority match
	sortedSettings := a.sortKeywordTriggersByPriority(matchingSettings)
	return &sortedSettings[0], nil
}

// matchesStoryID checks if the incoming story ID matches any of the configured story IDs
func (a *AutoReplyChannelSettingAggregate) matchesStoryID(incomingStoryID string, configuredStoryIDs []string) bool {
	for _, configuredID := range configuredStoryIDs {
		if incomingStoryID == configuredID {
			return true
		}
	}
	return false
}

// validateGeneralTimeTrigger checks if the event matches any general time-based trigger settings
func (a *AutoReplyChannelSettingAggregate) validateGeneralTimeTrigger(event WebhookEvent) (*AutoReplyTriggerSetting, error) {
	var matchingSettings []AutoReplyTriggerSetting

	// Find all general time triggers that match the current time
	for _, setting := range a.TriggerSettings {
		if !setting.IsActive() {
			continue
		}

		if setting.IsGeneralTimeTrigger() {
			// Check if the current time matches the trigger schedule
			isMatch, err := a.isWithinSchedule(event.Timestamp, setting.TriggerScheduleType, setting.TriggerScheduleSettings)
			if err != nil {
				return nil, err
			}

			if isMatch {
				matchingSettings = append(matchingSettings, setting)
			}
		}
	}

	if len(matchingSettings) == 0 {
		return nil, nil // No matching general time triggers
	}

	// Sort by priority and return the highest priority match
	sortedSettings := a.sortGeneralTriggersByPriority(matchingSettings)
	return &sortedSettings[0], nil
}

// validateIGStoryGeneralTrigger checks if the event matches any IG Story general time-based trigger settings
func (a *AutoReplyChannelSettingAggregate) validateIGStoryGeneralTrigger(event WebhookEvent) (*AutoReplyTriggerSetting, error) {
	// Must have IG Story ID
	if event.IGStoryID == nil {
		return nil, nil
	}

	var matchingSettings []AutoReplyTriggerSetting

	// Find all IG Story general time triggers that match the current time and story ID
	for _, setting := range a.TriggerSettings {
		if !setting.IsActive() {
			continue
		}

		if setting.IsIGStoryGeneralTrigger() {
			// Check if the story ID matches
			if !a.matchesStoryID(*event.IGStoryID, setting.IGStoryIDs) {
				continue
			}

			// Check if the current time matches the trigger schedule
			isMatch, err := a.isWithinSchedule(event.Timestamp, setting.TriggerScheduleType, setting.TriggerScheduleSettings)
			if err != nil {
				return nil, err
			}

			if isMatch {
				matchingSettings = append(matchingSettings, setting)
			}
		}
	}

	if len(matchingSettings) == 0 {
		return nil, nil // No matching IG Story general time triggers
	}

	// Sort by priority and return the highest priority match
	sortedSettings := a.sortGeneralTriggersByPriority(matchingSettings)
	return &sortedSettings[0], nil
}

// isWithinSchedule checks if the given timestamp is within the trigger schedule
func (a *AutoReplyChannelSettingAggregate) isWithinSchedule(timestamp time.Time, scheduleType *WebhookTriggerScheduleType, scheduleSettings json.RawMessage) (bool, error) {
	if scheduleType == nil {
		return false, nil
	}

	// Convert timestamp to the configured timezone
	loc, err := time.LoadLocation(a.Timezone)
	if err != nil {
		// Fallback to UTC if timezone is invalid
		loc = time.UTC
	}
	localTime := timestamp.In(loc)

	switch *scheduleType {
	case WebhookTriggerScheduleTypeDaily:
		return a.isWithinDailySchedule(localTime, scheduleSettings)
	case WebhookTriggerScheduleTypeMonthly:
		return a.isWithinMonthlySchedule(localTime, scheduleSettings)
	case WebhookTriggerScheduleTypeBusinessHour:
		return a.isWithinBusinessHours(localTime), nil
	case WebhookTriggerScheduleTypeNonBusinessHour:
		return !a.isWithinBusinessHours(localTime), nil
	default:
		return false, nil
	}
}

// isWithinDailySchedule checks if the time is within daily schedule settings
func (a *AutoReplyChannelSettingAggregate) isWithinDailySchedule(localTime time.Time, scheduleSettings json.RawMessage) (bool, error) {
	if len(scheduleSettings) == 0 {
		return false, nil
	}

	var dailySchedules []struct {
		StartTime string `json:"start_time"`
		EndTime   string `json:"end_time"`
	}

	if err := json.Unmarshal(scheduleSettings, &dailySchedules); err != nil {
		return false, err
	}

	currentTime := localTime.Format("15:04")

	for _, schedule := range dailySchedules {
		if a.isTimeInRange(currentTime, schedule.StartTime, schedule.EndTime) {
			return true, nil
		}
	}

	return false, nil
}

// isWithinMonthlySchedule checks if the time is within monthly schedule settings
func (a *AutoReplyChannelSettingAggregate) isWithinMonthlySchedule(localTime time.Time, scheduleSettings json.RawMessage) (bool, error) {
	if len(scheduleSettings) == 0 {
		return false, nil
	}

	var monthlySchedules []struct {
		Day       int    `json:"day"`
		StartTime string `json:"start_time"`
		EndTime   string `json:"end_time"`
	}

	if err := json.Unmarshal(scheduleSettings, &monthlySchedules); err != nil {
		return false, err
	}

	currentDay := localTime.Day()
	currentTime := localTime.Format("15:04")

	for _, schedule := range monthlySchedules {
		if currentDay == schedule.Day && a.isTimeInRange(currentTime, schedule.StartTime, schedule.EndTime) {
			return true, nil
		}
	}

	return false, nil
}

// isWithinBusinessHours checks if the given timestamp is within business hours
func (a *AutoReplyChannelSettingAggregate) isWithinBusinessHours(timestamp time.Time) bool {
	currentWeekday := int(timestamp.Weekday())
	// Convert Go's Sunday=0 to our Monday=1, Sunday=7 format
	if currentWeekday == 0 {
		currentWeekday = 7
	}

	currentTime := timestamp.Format("15:04")

	for _, businessHour := range a.BusinessHours {
		if int(businessHour.Weekday) == currentWeekday {
			startTime := businessHour.StartTime.Format("15:04")
			endTime := businessHour.EndTime.Format("15:04")
			if a.isTimeInRange(currentTime, startTime, endTime) {
				return true
			}
		}
	}

	return false
}

// isTimeInRange checks if currentTime is within the range [startTime, endTime)
// Handles midnight crossing (e.g., 22:00 to 06:00)
func (a *AutoReplyChannelSettingAggregate) isTimeInRange(currentTime, startTime, endTime string) bool {
	// Parse times
	current, err := time.Parse("15:04", currentTime)
	if err != nil {
		return false
	}
	start, err := time.Parse("15:04", startTime)
	if err != nil {
		return false
	}
	end, err := time.Parse("15:04", endTime)
	if err != nil {
		return false
	}

	// Check if range crosses midnight
	if start.After(end) {
		// Midnight crossing: start_time <= current_time OR current_time < end_time
		return !current.Before(start) || current.Before(end)
	} else {
		// Normal range: start_time <= current_time < end_time
		return !current.Before(start) && current.Before(end)
	}
}

// sortKeywordTriggersByPriority sorts keyword trigger settings by auto-reply priority
// Higher priority number = higher priority (e.g., priority 10 > priority 5)
func (a *AutoReplyChannelSettingAggregate) sortKeywordTriggersByPriority(settings []AutoReplyTriggerSetting) []AutoReplyTriggerSetting {
	// Make a copy to avoid modifying the original slice
	sortedSettings := make([]AutoReplyTriggerSetting, len(settings))
	copy(sortedSettings, settings)

	// Sort by AutoReplyPriority in descending order (higher number = higher priority)
	sort.Slice(sortedSettings, func(i, j int) bool {
		return sortedSettings[i].AutoReplyPriority > sortedSettings[j].AutoReplyPriority
	})

	return sortedSettings
}

// sortGeneralTriggersByPriority sorts general time trigger settings by:
// 1. Schedule type priority (monthly -> business hour -> non-business hour -> daily)
// 2. Auto-reply priority within the same schedule type (higher number = higher priority)
func (a *AutoReplyChannelSettingAggregate) sortGeneralTriggersByPriority(settings []AutoReplyTriggerSetting) []AutoReplyTriggerSetting {
	// Make a copy to avoid modifying the original slice
	sortedSettings := make([]AutoReplyTriggerSetting, len(settings))
	copy(sortedSettings, settings)

	// Sort by schedule type priority first, then by auto-reply priority
	sort.Slice(sortedSettings, func(i, j int) bool {
		scheduleTypePriorityI := getScheduleTypePriority(sortedSettings[i].TriggerScheduleType)
		scheduleTypePriorityJ := getScheduleTypePriority(sortedSettings[j].TriggerScheduleType)

		// If schedule type priorities are different, sort by schedule type priority
		if scheduleTypePriorityI != scheduleTypePriorityJ {
			return scheduleTypePriorityI < scheduleTypePriorityJ // Lower number = higher priority for schedule type
		}

		// If schedule type priorities are the same, sort by auto-reply priority
		return sortedSettings[i].AutoReplyPriority > sortedSettings[j].AutoReplyPriority // Higher number = higher priority
	})

	return sortedSettings
}

// getScheduleTypePriority returns the priority order for general time trigger schedule types
// Lower number = higher priority
func getScheduleTypePriority(scheduleType *WebhookTriggerScheduleType) int {
	if scheduleType == nil {
		return 999 // Lowest priority for nil schedule type
	}

	switch *scheduleType {
	case WebhookTriggerScheduleTypeMonthly:
		return 1 // Highest priority
	case WebhookTriggerScheduleTypeBusinessHour:
		return 2
	case WebhookTriggerScheduleTypeNonBusinessHour:
		return 3
	case WebhookTriggerScheduleTypeDaily:
		return 4 // Lowest priority
	default:
		return 999 // Unknown schedule type
	}
}

File: python_src/internal/domain/auto_reply/trigger_validation.py
"""Auto-reply trigger validation domain models and logic."""

from datetime import datetime, time
from typing import Optional

import pytz
from pydantic import BaseModel, Field

from internal.domain.auto_reply.auto_reply import AutoReply, AutoReplyEventType, AutoReplyStatus
from internal.domain.auto_reply.webhook_trigger import (
    MessageEvent,
    WebhookTriggerEventType,
    WebhookTriggerScheduleSettings,
    WebhookTriggerScheduleType,
    WebhookTriggerSetting,
)
from internal.domain.organization.business_hour_checker import BusinessHourChecker


class AutoReplyTriggerSetting(BaseModel):
    """Merged domain model combining AutoReply and WebhookTriggerSetting.

    This model provides unified access to both auto-reply configuration
    and webhook trigger settings for validation logic.
    """

    # AutoReply fields
    auto_reply_id: int = Field(..., description="Auto reply ID")
    auto_reply_name: str = Field(..., description="Auto reply name")
    auto_reply_status: AutoReplyStatus = Field(..., description="Auto reply status")
    auto_reply_event_type: AutoReplyEventType = Field(..., description="Auto reply event type")
    auto_reply_priority: int = Field(..., description="Auto reply priority for sorting")
    keywords: list[str] | None = Field(None, description="Keywords for keyword triggers")
    ig_story_ids: list[str] | None = Field(None, description="Instagram Story IDs for IG Story-specific triggers")

    # WebhookTriggerSetting fields
    webhook_trigger_id: int = Field(..., description="Webhook trigger setting ID")
    bot_id: int = Field(..., description="Bot ID")
    enable: bool = Field(..., description="Whether trigger is enabled")
    webhook_event_type: WebhookTriggerEventType = Field(..., description="Webhook event type")
    trigger_schedule_type: WebhookTriggerScheduleType | None = Field(None, description="Schedule type")
    trigger_schedule_settings: WebhookTriggerScheduleSettings | None = Field(None, description="Schedule settings")
    created_at: datetime = Field(..., description="Created timestamp")
    updated_at: datetime = Field(..., description="Updated timestamp")
    archived: bool = Field(False, description="Whether trigger is archived")

    class Config:
        """Pydantic configuration."""

        use_enum_values = True

    def is_active(self) -> bool:
        """Check if the trigger setting is active."""
        return self.enable and not self.archived and self.auto_reply_status == AutoReplyStatus.ACTIVE

    def is_keyword_trigger(self) -> bool:
        """Check if this is a keyword trigger."""
        return (
            self.auto_reply_event_type == AutoReplyEventType.KEYWORD
            and self.keywords is not None
            and len(self.keywords) > 0
        )

    def is_general_time_trigger(self) -> bool:
        """Check if this is a general time-based trigger."""
        return self.auto_reply_event_type == AutoReplyEventType.TIME and self.trigger_schedule_type is not None

    def is_ig_story_trigger(self) -> bool:
        """Check if this is an IG Story-specific trigger."""
        return self.ig_story_ids is not None and len(self.ig_story_ids) > 0

    def is_ig_story_keyword_trigger(self) -> bool:
        """Check if this is an IG Story keyword trigger."""
        return self.is_ig_story_trigger() and self.is_keyword_trigger()

    def is_ig_story_general_trigger(self) -> bool:
        """Check if this is an IG Story general (time-based) trigger."""
        return self.is_ig_story_trigger() and self.is_general_time_trigger()

    def is_general_keyword_trigger(self) -> bool:
        """Check if this is a general keyword trigger (not IG Story-specific)."""
        return self.is_keyword_trigger() and not self.is_ig_story_trigger()

    def is_general_time_only_trigger(self) -> bool:
        """Check if this is a general time-based trigger (not IG Story-specific)."""
        return self.is_general_time_trigger() and not self.is_ig_story_trigger()

    def normalize_keyword(self, keyword: str) -> str:
        """Normalize keyword for matching: case insensitive + trim spaces."""
        return keyword.strip().lower()

    def matches_keyword(self, message_content: str) -> bool:
        """Check if message content matches any configured keyword.

        Matching rules:
        - Case insensitive
        - Trim spaces
        - Exact match (not substring)
        """
        if not self.is_keyword_trigger():
            return False

        normalized_content = self.normalize_keyword(message_content)

        for keyword in self.keywords:
            normalized_keyword = self.normalize_keyword(keyword)
            if normalized_content == normalized_keyword:
                return True

        return False

    def matches_ig_story_id(self, ig_story_id: str | None) -> bool:
        """Check if the provided IG Story ID matches any configured story IDs."""
        if not self.is_ig_story_trigger() or ig_story_id is None:
            return False
        
        return ig_story_id in self.ig_story_ids

    @classmethod
    def from_auto_reply_and_webhook_trigger(
        cls, auto_reply: AutoReply, webhook_trigger: WebhookTriggerSetting
    ) -> "AutoReplyTriggerSetting":
        """Create AutoReplyTriggerSetting from AutoReply and WebhookTriggerSetting."""
        return cls(
            auto_reply_id=auto_reply.id,
            auto_reply_name=auto_reply.name,
            auto_reply_status=auto_reply.status,
            auto_reply_event_type=auto_reply.event_type,
            auto_reply_priority=auto_reply.priority,
            keywords=auto_reply.keywords,
            ig_story_ids=getattr(auto_reply, 'ig_story_ids', None),
            webhook_trigger_id=webhook_trigger.id,
            bot_id=webhook_trigger.bot_id,
            enable=webhook_trigger.enable,
            webhook_event_type=webhook_trigger.event_type,
            trigger_schedule_type=webhook_trigger.trigger_schedule_type,
            trigger_schedule_settings=webhook_trigger.trigger_schedule_settings,
            created_at=webhook_trigger.created_at,
            updated_at=webhook_trigger.updated_at,
            archived=webhook_trigger.archived,
        )


class AutoReplyChannelSettingAggregate(BaseModel):
    """Aggregate for auto-reply channel settings and validation logic."""

    bot_id: int = Field(..., description="Bot/Channel ID")
    organization_id: int = Field(..., description="Organization ID")
    timezone: str = Field(..., description="Timezone for time-based calculations")
    trigger_settings: list[AutoReplyTriggerSetting] = Field(..., description="All trigger settings for this bot")
    business_hour_checker: Optional[BusinessHourChecker] = Field(None, description="Business hour checker")

    class Config:
        """Pydantic configuration."""

        arbitrary_types_allowed = True

    def validate_trigger(self, event: MessageEvent) -> Optional[AutoReplyTriggerSetting]:
        """Validate trigger and return the first matching trigger setting based on priority.

        Priority system (PRD part2 - 4 levels):
        1. IG Story Keyword (highest priority)
        2. IG Story General (time-based)
        3. General Keyword
        4. General Time-based (lowest priority)

        Within each trigger type, sort by auto_reply_priority (higher number = higher priority).

        Args:
            event: The message event to validate

        Returns:
            The first matching AutoReplyTriggerSetting or None if no match
        """
        # Filter active triggers
        active_triggers = [trigger for trigger in self.trigger_settings if trigger.is_active()]

        if not active_triggers:
            return None

        # Convert event timestamp to bot timezone
        tz = pytz.timezone(self.timezone)
        event_time = event.timestamp.astimezone(tz)

        # Separate triggers by type (4-level priority system)
        ig_story_keyword_triggers = [trigger for trigger in active_triggers if trigger.is_ig_story_keyword_trigger()]
        ig_story_general_triggers = [trigger for trigger in active_triggers if trigger.is_ig_story_general_trigger()]
        general_keyword_triggers = [trigger for trigger in active_triggers if trigger.is_general_keyword_trigger()]
        general_time_triggers = [trigger for trigger in active_triggers if trigger.is_general_time_only_trigger()]

        # Sort triggers by priority (higher number = higher priority)
        ig_story_keyword_triggers.sort(key=lambda x: x.auto_reply_priority, reverse=True)
        ig_story_general_triggers.sort(key=lambda x: x.auto_reply_priority, reverse=True)
        general_keyword_triggers.sort(key=lambda x: x.auto_reply_priority, reverse=True)
        general_time_triggers.sort(key=lambda x: x.auto_reply_priority, reverse=True)

        # Priority 1: IG Story Keyword triggers (highest priority)
        for trigger in ig_story_keyword_triggers:
            if trigger.matches_keyword(event.content) and trigger.matches_ig_story_id(event.ig_story_id):
                return trigger

        # Priority 2: IG Story General triggers
        for trigger in ig_story_general_triggers:
            if self._matches_time_trigger(trigger, event_time) and trigger.matches_ig_story_id(event.ig_story_id):
                return trigger

        # Priority 3: General Keyword triggers
        for trigger in general_keyword_triggers:
            if trigger.matches_keyword(event.content):
                return trigger

        # Priority 4: General Time-based triggers (lowest priority)
        for trigger in general_time_triggers:
            if self._matches_time_trigger(trigger, event_time):
                return trigger

        return None

    def _matches_time_trigger(self, trigger: AutoReplyTriggerSetting, event_time: datetime) -> bool:
        """Check if a time-based trigger matches the given event time."""
        if not trigger.is_general_time_trigger():
            return False

        schedule_type = trigger.trigger_schedule_type
        schedule_settings = trigger.trigger_schedule_settings

        if schedule_type == WebhookTriggerScheduleType.MONTHLY:
            return self._matches_monthly_schedule(schedule_settings, event_time)
        elif schedule_type == WebhookTriggerScheduleType.BUSINESS_HOUR:
            return self._matches_business_hour_schedule(event_time)
        elif schedule_type == WebhookTriggerScheduleType.NON_BUSINESS_HOUR:
            return self._matches_non_business_hour_schedule(event_time)
        elif schedule_type == WebhookTriggerScheduleType.DAILY:
            return self._matches_daily_schedule(schedule_settings, event_time)

        return False

    def _matches_monthly_schedule(
        self, schedule_settings: WebhookTriggerScheduleSettings, event_time: datetime
    ) -> bool:
        """Check if event time matches monthly schedule."""
        if not schedule_settings or not schedule_settings.schedules:
            return False

        for schedule in schedule_settings.schedules:
            if schedule.get_schedule_type() != WebhookTriggerScheduleType.MONTHLY:
                continue

            settings = schedule.get_schedule_settings()
            if not settings:
                continue

            # Check if day matches
            if event_time.day != settings.get("day"):
                continue

            # Check if time is within range
            start_time = time.fromisoformat(settings["start_time"]).replace(
                second=time.min.second, microsecond=time.min.microsecond
            )
            end_time = time.fromisoformat(settings["end_time"]).replace(
                second=time.max.second, microsecond=time.max.microsecond
            )

            # Handle normal time range
            if start_time <= end_time:
                if start_time <= event_time.time() < end_time:
                    return True
            else:
                # Handle midnight crossing (e.g., 22:00 to 06:00)
                if start_time <= event_time.time() or event_time.time() < end_time:
                    return True

        return False

    def _matches_business_hour_schedule(self, event_time: datetime) -> bool:
        """Check if event time is within business hours."""
        if not self.business_hour_checker:
            return False

        return self.business_hour_checker.is_in_business_hours(event_time)

    def _matches_non_business_hour_schedule(self, event_time: datetime) -> bool:
        """Check if event time is outside business hours."""
        if not self.business_hour_checker:
            return False

        return self.business_hour_checker.is_in_non_business_hours(event_time)

    def _matches_daily_schedule(self, schedule_settings: WebhookTriggerScheduleSettings, event_time: datetime) -> bool:
        """Check if event time matches daily schedule."""
        if not schedule_settings or not schedule_settings.schedules:
            return False

        for schedule in schedule_settings.schedules:
            if schedule.get_schedule_type() != WebhookTriggerScheduleType.DAILY:
                continue

            settings = schedule.get_schedule_settings()
            if not settings:
                continue

            start_time = time.fromisoformat(settings["start_time"]).replace(
                second=time.min.second, microsecond=time.min.microsecond
            )
            end_time = time.fromisoformat(settings["end_time"]).replace(
                second=time.max.second, microsecond=time.max.microsecond
            )

            # Handle normal time range
            if start_time <= end_time:
                if start_time <= event_time.time() < end_time:
                    return True
            else:
                # Handle midnight crossing (e.g., 22:00 to 06:00)
                if start_time <= event_time.time() or event_time.time() < end_time:
                    return True

        return False
